sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.cs

```diff
@@ -0,0 +1,227 @@
+        if (perCallIndex > perTryIndex) throw new ArgumentOutOfRangeException(nameof(perCallIndex), "perCallIndex cannot be greater than perTryIndex.");
+
+        if (policies.Span[policies.Length - 1] is not PipelineTransport)
+        {
+            throw new ArgumentException("Last policy in the array must be of type 'PipelineTransport'.", nameof(policies));
+        }
+
+        _transport = (PipelineTransport)policies.Span[policies.Length - 1];
+        _policies = policies;
+
+        _perCallIndex = perCallIndex;
+        _perTryIndex = perTryIndex;
+        _beforeTransportIndex = beforeTransportIndex;
+    }
+
+    public static ClientPipeline Create()
+        => Create(ClientPipelineOptions.Default,
+            ReadOnlySpan<PipelinePolicy>.Empty,
+            ReadOnlySpan<PipelinePolicy>.Empty,
+            ReadOnlySpan<PipelinePolicy>.Empty);
+
+    public static ClientPipeline Create(ClientPipelineOptions options, params PipelinePolicy[] perCallPolicies)
+        => Create(options,
+            perCallPolicies,
+            ReadOnlySpan<PipelinePolicy>.Empty,
+            ReadOnlySpan<PipelinePolicy>.Empty);
+
+    public static ClientPipeline Create(
+        ClientPipelineOptions options,
+        ReadOnlySpan<PipelinePolicy> perCallPolicies,
+        ReadOnlySpan<PipelinePolicy> perTryPolicies,
+        ReadOnlySpan<PipelinePolicy> beforeTransportPolicies)
+    {
+        Argument.AssertNotNull(options, nameof(options));
+
+        int pipelineLength = perCallPolicies.Length + perTryPolicies.Length + beforeTransportPolicies.Length;
+
+        if (options.PerTryPolicies != null)
+        {
+            pipelineLength += options.PerTryPolicies.Length;
+        }
+
+        if (options.PerCallPolicies != null)
+        {
+            pipelineLength += options.PerCallPolicies.Length;
+        }
+
+        if (options.BeforeTransportPolicies != null)
+        {
+            pipelineLength += options.BeforeTransportPolicies.Length;
+        }
+
+        // TODO: Retry and buffering policies will come in a later PR.
+        //pipelineLength++; // for retry policy
+        //pipelineLength++; // for response buffering policy
+        pipelineLength++; // for transport
+
+        PipelinePolicy[] policies = new PipelinePolicy[pipelineLength];
+
+        int index = 0;
+
+        perCallPolicies.CopyTo(policies.AsSpan(index));
+        index += perCallPolicies.Length;
+
+        if (options.PerCallPolicies != null)
+        {
+            options.PerCallPolicies.CopyTo(policies.AsSpan(index));
+            index += options.PerCallPolicies.Length;
+        }
+
+        int perCallIndex = index;
+
+        // TODO: RetryPolicy will come in a later PR.
+        //if (options.RetryPolicy != null)
+        //{
+        //    policies[index++] = options.RetryPolicy;
+        //}
+        //else
+        //{
+        //    policies[index++] = new RequestRetryPolicy();
+        //}
+
+        perTryPolicies.CopyTo(policies.AsSpan(index));
```

Comment:
These represent the library perRetryPolicies, whereas the options.PerRetryPolicies are customer provided?

Reply:
That is correct.  I know @KrzysztofCwalina isn't super wild about this separation and having both on options and in the constructor, but this pattern allows us to observe precedence rules where it will show up in other places - i.e. end-user settings are provided in options and client-author settings are passed as parameters to methods.  An alternative proposal was to put them all on options, but in some cases, like the message classifier and IIRC there are others, that will mean we lose track of precedence or potentially clobber an end-user setting.  I can try to call these out when they show up in later PRs, and I'm open to proposals for other designs that might solve the problem better.

Reply:
Stashing this eye-chart for later - at some point we'll want to go through the various ways of setting options and corresponding precedence rules with the Core v-team, but that's probably a topic for a later PR! https://github.com/Azure/azure-sdk-for-net/pull/40097#issue-1999868531